shared/src/main/scala/weaponregex/model/regextree/RegexTree.scala

```diff
@@ -3,18 +3,24 @@ package weaponregex.model.regextree
 import weaponregex.model.Location
 
 trait RegexTree {
+  // Won't only a Node have children, and a Leaf never? This feels like it would make more sense in the Node trait
```

Comment:
This allows us to use this as an abstraction wihtout needing to check if a token is a `Node` or a `Leaf`.

Reply:
I would then suggest creating something like a `foldLeft` on RegexTree to build a new tree if you don't want to check the subtype of the tree.